File: bench/suite/src/st/pidigits/javascript/pidigits.js
// pidigits: single-threaded, plain JavaScript with BigInt (Benchmarks Game
// step-by-step unbounded spigot, as in the reference C/GMP program).
'use strict';
const n = +process.argv[2] || 30;
let acc = 0n, den = 1n, num = 1n;

function extractDigit(nth) { return (num * nth + acc) / den; }

let out = '', line = '', i = 0, k = 0;
while (i < n) {
  k++;
  const k2 = BigInt(k * 2 + 1);
  acc = (acc + num * 2n) * k2;
  den *= k2;
  num *= BigInt(k);
  if (num > acc) continue;
  const d = extractDigit(3n);
  if (d !== extractDigit(4n)) continue;
  line += d.toString();
  i++;
  if (i % 10 === 0) { out += line + '\t:' + i + '\n'; line = ''; }
  acc = (acc - den * d) * 10n;
  num *= 10n;
}
if (line.length > 0) out += line.padEnd(10, ' ') + '\t:' + i + '\n';
process.stdout.write(out);

File: bench/suite/src/st/pidigits/python/pidigits.py
# pidigits: single-threaded, pure CPython int (Benchmarks Game step-by-step
# unbounded spigot, as in the reference C/GMP program).
import sys


def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 30
    acc, den, num = 0, 1, 1
    out = []
    line = []
    i = k = 0
    while i < n:
        k += 1
        k2 = k * 2 + 1
        acc = (acc + num * 2) * k2
        den *= k2
        num *= k
        if num > acc:
            continue
        d = (num * 3 + acc) // den
        if d != (num * 4 + acc) // den:
            continue
        line.append(chr(48 + d))
        i += 1
        if i % 10 == 0:
            out.append("%s\t:%d\n" % ("".join(line), i))
            line = []
        acc = (acc - den * d) * 10
        num *= 10
    if line:
        out.append("%s\t:%d\n" % ("".join(line).ljust(10), i))
    sys.stdout.write("".join(out))


main()
